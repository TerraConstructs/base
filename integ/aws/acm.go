package aws

import (
	"fmt"
	"time"

	"github.com/aws/aws-sdk-go/service/acm"
	"github.com/stretchr/testify/require"

	"github.com/gruntwork-io/terratest/modules/aws"
	"github.com/gruntwork-io/terratest/modules/logger"
	"github.com/gruntwork-io/terratest/modules/retry"
	"github.com/gruntwork-io/terratest/modules/testing"
)

// ref: https://github.com/gruntwork-io/terratest/blob/v0.47.1/modules/aws/acm.go

// Get Certificate Status
func GetAcmCertificateStatus(t testing.TestingT, awsRegion string, certArn string) string {
	status, err := GetAcmCertificateStatusE(t, awsRegion, certArn)
	if err != nil {
		t.Fatal(err)
	}
	return status
}

// GetAcmCertificateStatusE gets the ACM certificate status for the given certificate ARN in the given region.
func GetAcmCertificateStatusE(t testing.TestingT, awsRegion string, certArn string) (string, error) {
	acmClient, err := aws.NewAcmClientE(t, awsRegion)
	if err != nil {
		return "", err
	}

	result, err := acmClient.DescribeCertificate(&acm.DescribeCertificateInput{
		CertificateArn: &certArn,
	})
	if err != nil {
		return "", err
	}

	return *result.Certificate.Status, nil
}

// WaitForCertificateIssued waits for the certificate to be issued
func WaitForCertificateIssued(
	t testing.TestingT,
	certArn string,
	region string,
	maxRetries int,
	sleepBetweenRetries time.Duration,
) {
	err := WaitForCertificateIssuedE(t, certArn, region, maxRetries, sleepBetweenRetries)
	require.NoError(t, err)
}

// WaitForCertificateIssuedE waits for the ACM Certificate to be issued
func WaitForCertificateIssuedE(
	t testing.TestingT,
	certArn string,
	region string,
	maxRetries int,
	sleepBetweenRetries time.Duration,
) error {
	msg, err := retry.DoWithRetryE(
		t,
		fmt.Sprintf("Waiting for Certificate %s to be ISSUED.", certArn),
		maxRetries,
		sleepBetweenRetries,
		func() (string, error) {
			certStatus, err := GetAcmCertificateStatusE(t, region, certArn)
			if err != nil {
				return "", err
			}
			if certStatus != acm.CertificateStatusIssued {
				return "", NewCertificateNotIssuedError(certArn, certStatus)
			}
			return fmt.Sprintf("Certificate %s is now at desired status %s", certArn, acm.CertificateStatusIssued), nil
		},
	)
	logger.Log(t, msg)
	return err
}
