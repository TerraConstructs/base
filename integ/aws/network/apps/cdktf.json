{
  "//": "Dummy file to satisfy the bundling findUp logic"
}
