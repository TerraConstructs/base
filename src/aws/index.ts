// core
// export * from "./tag-aspect";
// export * from "./tag-manager";
export * from "./arn";
export * from "./aws-construct";
export * from "./aws-stack";
export * from "./aws-tags";
export * from "./log-retention";

// core JSII compatible generated interfaces
export * from "./provider-config.generated";

// classes
export * as iam from "./iam";
export * as encryption from "./encryption";
export * as compute from "./compute";
export * as network from "./network";
export * as storage from "./storage";
export * as notify from "./notify";
export * as edge from "./edge";
export * as cloudwatch from "./cloudwatch";
