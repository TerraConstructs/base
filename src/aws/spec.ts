import {
  dataAwsAvailabilityZones,
  dataAwsCallerIdentity,
  dataAwsPartition,
  dataAwsRegion,
  dataAwsServicePrincipal,
  provider,
} from "@cdktf/provider-aws";
import {
  TerraformStack,
  TerraformIterator,
  Token,
  Fn,
  ResourceTerraformIterator,
} from "cdktf";
import { snakeCase } from "change-case";
import { Construct, IConstruct } from "constructs";
import { Arn, ArnComponents, ArnFormat, AwsProviderConfig } from ".";
import { SpecBaseProps, SpecBase, ISpec } from "../";

const AWS_SPEC_SYMBOL = Symbol.for("@envtio/base/lib/aws.AwsSpec");

export interface AwsSpecProps extends SpecBaseProps {
  /**
   * The AWS Provider configuration (without the alias field)
   */
  readonly providerConfig: AwsProviderConfig;
}

export interface IAwsSpec extends ISpec {
  /**
   * The AWS Region for the beacon
   */
  readonly region: string;
  /**
   * The AWS Account for the beacon
   */
  readonly account: string;
  /**
   * The AWS Partition for the beacon
   */
  readonly partition: string;

  /**
   * The service Principal Id for a specific service
   *
   * @param serviceName The service name to get the service principal ID for
   * @param region The region to get the service principal ID for
   */
  servicePrincipalName(serviceName: string, region?: string): string;
  // /**
  //  * Produce the Token's value at resolution time
  //  */
  // resolve<T>(obj: T): T;
}

interface AwsLookup {
  awsProvider: provider.AwsProvider;
  dataAwsRegion?: dataAwsRegion.DataAwsRegion;
  dataAwsCallerIdentity?: dataAwsCallerIdentity.DataAwsCallerIdentity;
  dataAwsPartition?: dataAwsPartition.DataAwsPartition;
  dataAwsAvailabilityZones?: dataAwsAvailabilityZones.DataAwsAvailabilityZones;
  // AWS Service Principals by region and by service
  dataAwsServicePrincipals: Record<
    string,
    Record<string, dataAwsServicePrincipal.DataAwsServicePrincipal>
  >;
}

/**
 * A Terraform stack constrained to a single AWS Account/Region to simulate CFN behavior.
 */
export class AwsSpec extends SpecBase implements IAwsSpec {
  // ref: https://github.com/aws/aws-cdk/blob/v2.150.0/packages/aws-cdk-lib/core/lib/stack.ts#L204

  /**
   * Return whether the given object is a Stack.
   *
   * attribute detection since as 'instanceof' potentially fails across Library releases.
   */
  public static isAwsSpec(x: any): x is AwsSpec {
    return x !== null && typeof x === "object" && AWS_SPEC_SYMBOL in x;
  }

  // ref: https://github.com/aws/aws-cdk/blob/v2.150.0/packages/aws-cdk-lib/core/lib/stack.ts#L212

  /**
   * Looks up the first stack scope in which `construct` is defined. Fails if there is no stack up the tree or the stack is not an AwsSpec.
   * @param construct The construct to start the search from.
   */
  public static ofAwsBeacon(construct: IConstruct): AwsSpec {
    const s = TerraformStack.of(construct);
    if (AwsSpec.isAwsSpec(s)) {
      return s;
    }
    throw new Error(
      `Resource '${construct.constructor?.name}' at '${construct.node.path}' should be created in the scope of an AwsSpec, but no AwsSpec found`,
    );
  }

  private readonly lookup: AwsLookup;
  private regionalAwsProviders: { [region: string]: provider.AwsProvider } = {};

  constructor(scope: Construct, id: string, props: AwsSpecProps) {
    super(scope, id, props);
    this.lookup = {
      awsProvider: new provider.AwsProvider(
        this,
        "defaultAwsProvider",
        props.providerConfig,
      ),
      dataAwsServicePrincipals: {},
    };
    Object.defineProperty(this, AWS_SPEC_SYMBOL, { value: true });
  }

  public get provider(): provider.AwsProvider {
    return this.lookup.awsProvider;
  }

  /**
   * Get the Region of the AWS Stack
   */
  public get region(): string {
    if (this.lookup.awsProvider.region) {
      return this.lookup.awsProvider.region;
    }
    if (!this.lookup.dataAwsRegion) {
      this.lookup.dataAwsRegion = new dataAwsRegion.DataAwsRegion(
        this,
        "Region",
        {
          provider: this.lookup.awsProvider,
        },
      );
    }
    return this.lookup.dataAwsRegion.name;
  }

  private get dataAwsCallerIdentity(): dataAwsCallerIdentity.DataAwsCallerIdentity {
    if (!this.lookup.dataAwsCallerIdentity) {
      this.lookup.dataAwsCallerIdentity =
        new dataAwsCallerIdentity.DataAwsCallerIdentity(
          this,
          "CallerIdentity",
          {
            provider: this.lookup.awsProvider,
          },
        );
    }
    return this.lookup.dataAwsCallerIdentity;
  }

  private get dataAwsAvailabilityZones(): dataAwsAvailabilityZones.DataAwsAvailabilityZones {
    if (!this.lookup.dataAwsAvailabilityZones) {
      this.lookup.dataAwsAvailabilityZones =
        new dataAwsAvailabilityZones.DataAwsAvailabilityZones(
          this,
          "AvailabilityZones",
          {
            provider: this.lookup.awsProvider,
          },
        );
    }
    return this.lookup.dataAwsAvailabilityZones;
  }

  private get dataAwsPartition(): dataAwsPartition.DataAwsPartition {
    if (!this.lookup.dataAwsPartition) {
      this.lookup.dataAwsPartition = new dataAwsPartition.DataAwsPartition(
        this,
        "Partitition",
        {
          provider: this.lookup.awsProvider,
        },
      );
    }
    return this.lookup.dataAwsPartition;
  }

  private getRegionalAwsProvider(region: string): provider.AwsProvider {
    if (!this.regionalAwsProviders[region]) {
      this.regionalAwsProviders[region] = new provider.AwsProvider(
        this,
        `aws_${toTerraformIdentifier(region)}`,
        {
          region,
          alias: toTerraformIdentifier(region),
        },
      );
    }
    return this.regionalAwsProviders[region];
  }

  /**
   * Get the Account of the AWS Stack
   */
  public get account(): string {
    return this.dataAwsCallerIdentity.accountId;
  }

  /**
   * Get the Partition of the AWS Stack
   */
  public get partition() {
    return this.dataAwsPartition.partition;
  }

  /**
   * Base DNS domain name for the current partition (e.g., amazonaws.com in AWS Commercial, amazonaws.com.cn in AWS China).
   */
  public get urlSuffix() {
    return this.dataAwsPartition.dnsSuffix;
  }

  /**
   * Return the service principal name based on the region it's used in.
   *
   * Some service principal names used to be different for different partitions,
   * and some were not.
   *
   * These days all service principal names are standardized, and they are all
   * of the form `<servicename>.amazonaws.com`.
   *
   * To avoid breaking changes, handling is provided for services added with the formats below,
   * however, no additional handling will be added for new regions or partitions.
   *   - s3
   *   - s3.amazonaws.com
   *   - s3.amazonaws.com.cn
   *   - s3.c2s.ic.gov
   *   - s3.sc2s.sgov.gov
   *
   * @param service The service name to get the service principal ID for
   * @param region The region to get the service principal ID for
   */
  public servicePrincipalName(service: string, region?: string): string {
    const DEFAULT_REGION_KEY = "default_region";
    if (!region) {
      region = DEFAULT_REGION_KEY;
    }

    if (Token.isUnresolved(region)) {
      throw new Error(
        "Cannot determine the service principal ID because the region is a token. " +
          "You must specify the region explicitly.",
      );
    }

    // if full service name is provided, extract just the service name
    // for supported cases (as required by Terraform aws_service_principal Data Source)
    const match = service.match(
      /^([^.]+)(?:(?:\.amazonaws\.com(?:\.cn)?)|(?:\.c2s\.ic\.gov)|(?:\.sc2s\.sgov\.gov))?$/,
    );
    const serviceName = match ? match[1] : service;
    if (!this.lookup.dataAwsServicePrincipals[region]) {
      this.lookup.dataAwsServicePrincipals[region] = {};
    }
    if (!this.lookup.dataAwsServicePrincipals[region][serviceName]) {
      this.lookup.dataAwsServicePrincipals[region][serviceName] =
        new dataAwsServicePrincipal.DataAwsServicePrincipal(
          this,
          `aws_svcp_${toTerraformIdentifier(region)}_${serviceName}}`,
          {
            serviceName,
            provider:
              region === DEFAULT_REGION_KEY
                ? undefined
                : this.getRegionalAwsProvider(region),
          },
        );
    }
    return this.lookup.dataAwsServicePrincipals[region][serviceName].name;
  }

  /**
   * Creates an ARN from components.
   *
   * If `partition`, `region` or `account` are not specified, the stack's
   * partition, region and account will be used.
   *
   * If any component is the empty string, an empty string will be inserted
   * into the generated ARN at the location that component corresponds to.
   *
   * The ARN will be formatted as follows:
   *
   *   arn:{partition}:{service}:{region}:{account}:{resource}{sep}{resource-name}
   *
   * The required ARN pieces that are omitted will be taken from the stack that
   * the 'scope' is attached to. If all ARN pieces are supplied, the supplied scope
   * can be 'undefined'.
   */
  public formatArn(components: ArnComponents): string {
    return Arn.format(components, this);
  }

  /**
   * Given an ARN, parses it and returns components.
   *
   * IF THE ARN IS A CONCRETE STRING...
   *
   * ...it will be parsed and validated. The separator (`sep`) will be set to '/'
   * if the 6th component includes a '/', in which case, `resource` will be set
   * to the value before the '/' and `resourceName` will be the rest. In case
   * there is no '/', `resource` will be set to the 6th components and
   * `resourceName` will be set to the rest of the string.
   *
   * IF THE ARN IS A TOKEN...
   *
   * ...it cannot be validated, since we don't have the actual value yet at the
   * time of this function call. You will have to supply `sepIfToken` and
   * whether or not ARNs of the expected format usually have resource names
   * in order to parse it properly. The resulting `ArnComponents` object will
   * contain tokens for the subexpressions of the ARN, not string literals.
   *
   * If the resource name could possibly contain the separator char, the actual
   * resource name cannot be properly parsed. This only occurs if the separator
   * char is '/', and happens for example for S3 object ARNs, IAM Role ARNs,
   * IAM OIDC Provider ARNs, etc. To properly extract the resource name from a
   * Tokenized ARN, you must know the resource type and call
   * `Arn.extractResourceName`.
   *
   * @param arn The ARN string to parse
   * @param sepIfToken The separator used to separate resource from resourceName
   * @param hasName Whether there is a name component in the ARN at all. For
   * example, SNS Topics ARNs have the 'resource' component contain the topic
   * name, and no 'resourceName' component.
   *
   * @returns an ArnComponents object which allows access to the various
   * components of the ARN.
   *
   * @returns an ArnComponents object which allows access to the various
   *      components of the ARN.
   *
   * @deprecated use splitArn instead
   */
  public parseArn(
    arn: string,
    sepIfToken: string = "/",
    hasName: boolean = true,
  ): ArnComponents {
    return Arn.parse(arn, sepIfToken, hasName);
  }

  /**
   * Splits the provided ARN into its components.
   * Works both if 'arn' is a string like 'arn:aws:s3:::bucket',
   * and a Token representing a dynamic CloudFormation expression
   * (in which case the returned components will also be dynamic CloudFormation expressions,
   * encoded as Tokens).
   *
   * @param arn the ARN to split into its components
   * @param arnFormat the expected format of 'arn' - depends on what format the service 'arn' represents uses
   */
  public splitArn(arn: string, arnFormat: ArnFormat): ArnComponents {
    return Arn.split(arn, arnFormat);
  }

  /**
   * Returns iterator for all AZs that are available in the AWS environment
   * (account/region) associated with this stack (default or aliased provider).
   *
   * this will return a cdktf iterator
   *
   * https://developer.hashicorp.com/terraform/cdktf/concepts/iterators#define-iterators
   *
   * To specify a different strategy for selecting availability zones override this method.
   */
  public get availabilityZoneIterator(): ResourceTerraformIterator {
    const azs = this.dataAwsAvailabilityZones;
    return TerraformIterator.fromDataSources(azs);
  }

  /**
   * Returns a Token as List of AZ names that are available in the stack's
   * AWS environment (account/region).
   *
   * The list is slized by `maxCount` which defaults to 2.
   *
   * Note: Must use `Fn.Index` to access the AZ names.
   *
   * @param maxCount the maximum number of AZs to return
   */
  public availabilityZones(maxCount: number = 2): string[] {
    const azs = this.dataAwsAvailabilityZones;
    return Token.asList(Fn.slice(azs.names, 0, maxCount));
  }

  // /**
  //  * Resolve a tokenized value in the context of the current stack.
  //  */
  // public resolve<T>(obj: T): T {
  //   // ref: https://github.com/hashicorp/terraform-cdk/blob/v0.20.7/packages/cdktf/lib/terraform-stack.ts#L151
  //   // ref: https://github.com/aws/aws-cdk/blob/v2.150.0/packages/aws-cdk-lib/core/lib/stack.ts#L572
  //   return resolve(this, obj);
  // }
}

function toTerraformIdentifier(identifier: string) {
  return snakeCase(identifier).replace(/-/g, "_");
}
