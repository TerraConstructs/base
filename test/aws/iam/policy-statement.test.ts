import { Testing } from "cdktf";
import "cdktf/lib/testing/adapters/jest";
import {
  AnyPrincipal,
  PolicyDocument,
  PolicyStatement,
  Effect,
} from "../../../src/aws/iam";
import { AwsSpec } from "../../../src/aws/spec";

const environmentName = "Test";
const gridUUID = "123e4567-e89b-12d3";
const providerConfig = { region: "us-east-1" };
describe("IAM policy statement", () => {
  describe("from JSON", () => {
    test("parses with no principal", () => {
      // GIVEN
      const spec1 = getAwsSpec();
      const spec2 = getAwsSpec();

      const s = new PolicyStatement();
      s.addActions("service:action1", "service:action2");
      s.addAllResources();
      s.addConditionObject("test", { key: "value" });

      const doc1 = new PolicyDocument(spec1, "doc");
      doc1.addStatements(s);

      // when
      PolicyDocument.fromJson(spec2, "doc", doc1.toDocumentJson());

      // then
      const doc1Synth = Testing.synth(spec1);
      expect(doc1Synth).toEqual(Testing.synth(spec2));
      expect(doc1Synth).toMatchSnapshot();
    });

    test("parses a given Principal", () => {
      const spec1 = getAwsSpec();
      const spec2 = getAwsSpec();

      const s = new PolicyStatement();
      s.addActions("service:action1", "service:action2");
      s.addAllResources();
      s.addArnPrincipal("somearn");
      s.addConditionObject("equals", { key: "value" });

      const doc1 = new PolicyDocument(spec1, "doc");
      doc1.addStatements(s);

      PolicyDocument.fromJson(spec2, "doc", doc1.toDocumentJson());

      const doc1Synth = Testing.synth(spec1);
      expect(doc1Synth).toEqual(Testing.synth(spec2));
      expect(doc1Synth).toMatchSnapshot();
    });

    test("should not convert `Principal: *` to `Principal: { AWS: * }`", () => {
      const spec = getAwsSpec();
      const s = PolicyStatement.fromJson({
        Action: ["service:action1"],
        Principal: "*",
        Resource: "*",
      });

      const doc1 = new PolicyDocument(spec, "doc");
      doc1.addStatements(s);

      const rendered = doc1.toDocumentJson();

      expect(rendered).toEqual({
        Statement: [
          {
            Action: "service:action1",
            Effect: "Allow",
            Principal: "*",
            Resource: "*",
          },
        ],
        Version: "2012-10-17",
      });
      expect(Testing.synth(spec)).toMatchSnapshot();
    });

    test("parses a given notPrincipal", () => {
      const spec1 = getAwsSpec();
      const spec2 = getAwsSpec();

      const s = new PolicyStatement();
      s.addActions("service:action1", "service:action2");
      s.addAllResources();
      s.addNotPrincipals(new AnyPrincipal());
      s.addConditionObject("equals", { key: "value" });

      const doc1 = new PolicyDocument(spec1, "doc");
      doc1.addStatements(s);

      PolicyDocument.fromJson(spec2, "doc", doc1.toDocumentJson());

      const doc1Synth = Testing.synth(spec1);
      expect(doc1Synth).toEqual(Testing.synth(spec2));
      expect(doc1Synth).toMatchSnapshot();
    });

    test("parses with notAction", () => {
      const spec1 = getAwsSpec();
      const spec2 = getAwsSpec();

      const s = new PolicyStatement();
      s.addNotActions("service:action3");
      s.addAllResources();

      const doc1 = new PolicyDocument(spec1, "doc");
      doc1.addStatements(s);

      PolicyDocument.fromJson(spec2, "doc", doc1.toDocumentJson());

      const doc1Synth = Testing.synth(spec1);
      expect(doc1Synth).toEqual(Testing.synth(spec2));
      expect(doc1Synth).toMatchSnapshot();
    });

    test("parses with notActions", () => {
      const spec1 = getAwsSpec();
      const spec2 = getAwsSpec();

      const s = new PolicyStatement();
      s.addNotActions("service:action3", "service:action4");
      s.addAllResources();

      const doc1 = new PolicyDocument(spec1, "doc");
      doc1.addStatements(s);

      PolicyDocument.fromJson(spec2, "doc", doc1.toDocumentJson());

      const doc1Synth = Testing.synth(spec1);
      expect(doc1Synth).toEqual(Testing.synth(spec2));
      expect(doc1Synth).toMatchSnapshot();
    });

    test("parses with notResource", () => {
      const spec1 = getAwsSpec();
      const spec2 = getAwsSpec();

      const s = new PolicyStatement();
      s.addActions("service:action3", "service:action4");
      s.addNotResources("resource1");

      const doc1 = new PolicyDocument(spec1, "doc");
      doc1.addStatements(s);

      PolicyDocument.fromJson(spec2, "doc", doc1.toDocumentJson());

      const doc1Synth = Testing.synth(spec1);
      expect(doc1Synth).toEqual(Testing.synth(spec2));
      expect(doc1Synth).toMatchSnapshot();
    });

    test("parses with notResources", () => {
      const spec1 = getAwsSpec();
      const spec2 = getAwsSpec();

      const s = new PolicyStatement();
      s.addActions("service:action3", "service:action4");
      s.addNotResources("resource1", "resource2");

      const doc1 = new PolicyDocument(spec1, "doc");
      doc1.addStatements(s);

      PolicyDocument.fromJson(spec2, "doc", doc1.toDocumentJson());

      const doc1Synth = Testing.synth(spec1);
      expect(doc1Synth).toEqual(Testing.synth(spec2));
      expect(doc1Synth).toMatchSnapshot();
    });

    test("the kitchen sink", () => {
      const spec = getAwsSpec();

      const policyDocument = {
        Version: "2012-10-17",
        Statement: [
          {
            Sid: "FirstStatement",
            Effect: "Allow",
            Action: "iam:ChangePassword",
            Resource: "*",
          },
          {
            Sid: "SecondStatement",
            Effect: "Allow",
            Action: "s3:ListAllMyBuckets",
            Resource: "*",
          },
          {
            Sid: "ThirdStatement",
            Effect: "Allow",
            Action: ["s3:List*", "s3:Get*"],
            Resource: [
              "arn:aws:s3:::confidential-data",
              "arn:aws:s3:::confidential-data/*",
            ],
            Condition: { Bool: { "aws:MultiFactorAuthPresent": "true" } },
          },
        ],
      };

      const doc = PolicyDocument.fromJson(spec, "doc", policyDocument);

      expect(doc.toDocumentJson()).toEqual(policyDocument);
      expect(Testing.synth(spec)).toMatchSnapshot();
    });

    test("throws error with field data being object", () => {
      expect(() => {
        PolicyStatement.fromJson({
          Action: {},
        });
      }).toThrow(/Fields must be either a string or an array of strings/);
    });

    test("throws error with field data being array of non-strings", () => {
      expect(() => {
        PolicyStatement.fromJson({
          Action: [{}],
        });
      }).toThrow(/Fields must be either a string or an array of strings/);
    });
  });

  // test("throws error when group is specified for 'Principal' or 'NotPrincipal'", () => {
  //   const spec = getAwsSpec();
  //   const group = new Group(spec, "groupId");
  //   const policyStatement = new PolicyStatement();

  //   expect(() => policyStatement.addPrincipals(group)).toThrow(
  //     /Cannot use an IAM Group as the 'Principal' or 'NotPrincipal' in an IAM Policy/,
  //   );
  //   expect(() => policyStatement.addNotPrincipals(group)).toThrow(
  //     /Cannot use an IAM Group as the 'Principal' or 'NotPrincipal' in an IAM Policy/,
  //   );
  // });

  test("throws error when an invalid 'Action' or 'NotAction' is added", () => {
    const policyStatement = new PolicyStatement();
    const invalidAction = "xyz";
    expect(() => policyStatement.addActions(invalidAction)).toThrow(
      `Action '${invalidAction}' is invalid. An action string consists of a service namespace, a colon, and the name of an action. Action names can include wildcards.`,
    );
    expect(() => policyStatement.addNotActions(invalidAction)).toThrow(
      `Action '${invalidAction}' is invalid. An action string consists of a service namespace, a colon, and the name of an action. Action names can include wildcards.`,
    );
  });

  test("multiple identical entries render to a scalar (instead of a singleton list)", () => {
    const policyStatement = new PolicyStatement({
      actions: ["aws:Action"],
    });

    policyStatement.addResources("asdf");
    policyStatement.addResources("asdf");
    policyStatement.addResources("asdf");

    expect(policyStatement.toStatementJson()).toEqual({
      Effect: "Allow",
      Action: "aws:Action",
      Resource: "asdf",
    });
  });

  test("a frozen policy statement cannot be modified any more", () => {
    // GIVEN
    const statement = new PolicyStatement({
      actions: ["action:a"],
      resources: ["*"],
    });
    statement.freeze();

    // WHEN
    const modifications = [
      () => (statement.sid = "asdf"),
      () => (statement.effect = Effect.DENY),
      () => statement.addActions("abc:def"),
      () => statement.addNotActions("abc:def"),
      () => statement.addResources("*"),
      () => statement.addNotResources("*"),
      () => statement.addPrincipals(new AnyPrincipal()),
      () => statement.addNotPrincipals(new AnyPrincipal()),
      () => statement.addConditionObject("equals", "value"),
    ];

    for (const mod of modifications) {
      expect(mod).toThrow(/can no longer be modified/);
    }
  });
});

function getAwsSpec(): AwsSpec {
  const app = Testing.app();
  return new AwsSpec(app, "TestSpec", {
    environmentName,
    gridUUID,
    providerConfig,
  });
}
