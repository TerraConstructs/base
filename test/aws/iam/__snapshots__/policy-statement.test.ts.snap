// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IAM policy statement from JSON parses a given Principal 1`] = `
"{
  "data": {
    "aws_caller_identity": {
      "CallerIdentity": {
        "provider": "aws"
      }
    },
    "aws_iam_policy_document": {
      "doc_138423F6": {
        "statement": [
          {
            "actions": [
              "service:action1",
              "service:action2"
            ],
            "condition": [
              {
                "test": "equals",
                "values": [
                  "value"
                ],
                "variable": "key"
              }
            ],
            "effect": "Allow",
            "principals": [
              {
                "identifiers": [
                  "somearn"
                ],
                "type": "AWS"
              }
            ],
            "resources": [
              "*"
            ]
          }
        ]
      }
    },
    "aws_partition": {
      "Partitition": {
        "provider": "aws"
      }
    }
  },
  "provider": {
    "aws": [
      {
        "region": "us-east-1"
      }
    ]
  },
  "terraform": {
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "5.68.0"
      }
    }
  }
}"
`;

exports[`IAM policy statement from JSON parses a given notPrincipal 1`] = `
"{
  "data": {
    "aws_caller_identity": {
      "CallerIdentity": {
        "provider": "aws"
      }
    },
    "aws_iam_policy_document": {
      "doc_138423F6": {
        "statement": [
          {
            "actions": [
              "service:action1",
              "service:action2"
            ],
            "condition": [
              {
                "test": "equals",
                "values": [
                  "value"
                ],
                "variable": "key"
              }
            ],
            "effect": "Allow",
            "not_principals": [
              {
                "identifiers": [
                  "*"
                ],
                "type": "AWS"
              }
            ],
            "resources": [
              "*"
            ]
          }
        ]
      }
    },
    "aws_partition": {
      "Partitition": {
        "provider": "aws"
      }
    }
  },
  "provider": {
    "aws": [
      {
        "region": "us-east-1"
      }
    ]
  },
  "terraform": {
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "5.68.0"
      }
    }
  }
}"
`;

exports[`IAM policy statement from JSON parses with no principal 1`] = `
"{
  "data": {
    "aws_caller_identity": {
      "CallerIdentity": {
        "provider": "aws"
      }
    },
    "aws_iam_policy_document": {
      "doc_138423F6": {
        "statement": [
          {
            "actions": [
              "service:action1",
              "service:action2"
            ],
            "condition": [
              {
                "test": "test",
                "values": [
                  "value"
                ],
                "variable": "key"
              }
            ],
            "effect": "Allow",
            "resources": [
              "*"
            ]
          }
        ]
      }
    },
    "aws_partition": {
      "Partitition": {
        "provider": "aws"
      }
    }
  },
  "provider": {
    "aws": [
      {
        "region": "us-east-1"
      }
    ]
  },
  "terraform": {
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "5.68.0"
      }
    }
  }
}"
`;

exports[`IAM policy statement from JSON parses with notAction 1`] = `
"{
  "data": {
    "aws_caller_identity": {
      "CallerIdentity": {
        "provider": "aws"
      }
    },
    "aws_iam_policy_document": {
      "doc_138423F6": {
        "statement": [
          {
            "effect": "Allow",
            "not_actions": [
              "service:action3"
            ],
            "resources": [
              "*"
            ]
          }
        ]
      }
    },
    "aws_partition": {
      "Partitition": {
        "provider": "aws"
      }
    }
  },
  "provider": {
    "aws": [
      {
        "region": "us-east-1"
      }
    ]
  },
  "terraform": {
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "5.68.0"
      }
    }
  }
}"
`;

exports[`IAM policy statement from JSON parses with notActions 1`] = `
"{
  "data": {
    "aws_caller_identity": {
      "CallerIdentity": {
        "provider": "aws"
      }
    },
    "aws_iam_policy_document": {
      "doc_138423F6": {
        "statement": [
          {
            "effect": "Allow",
            "not_actions": [
              "service:action3",
              "service:action4"
            ],
            "resources": [
              "*"
            ]
          }
        ]
      }
    },
    "aws_partition": {
      "Partitition": {
        "provider": "aws"
      }
    }
  },
  "provider": {
    "aws": [
      {
        "region": "us-east-1"
      }
    ]
  },
  "terraform": {
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "5.68.0"
      }
    }
  }
}"
`;

exports[`IAM policy statement from JSON parses with notResource 1`] = `
"{
  "data": {
    "aws_caller_identity": {
      "CallerIdentity": {
        "provider": "aws"
      }
    },
    "aws_iam_policy_document": {
      "doc_138423F6": {
        "statement": [
          {
            "actions": [
              "service:action3",
              "service:action4"
            ],
            "effect": "Allow",
            "not_resources": [
              "resource1"
            ]
          }
        ]
      }
    },
    "aws_partition": {
      "Partitition": {
        "provider": "aws"
      }
    }
  },
  "provider": {
    "aws": [
      {
        "region": "us-east-1"
      }
    ]
  },
  "terraform": {
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "5.68.0"
      }
    }
  }
}"
`;

exports[`IAM policy statement from JSON parses with notResources 1`] = `
"{
  "data": {
    "aws_caller_identity": {
      "CallerIdentity": {
        "provider": "aws"
      }
    },
    "aws_iam_policy_document": {
      "doc_138423F6": {
        "statement": [
          {
            "actions": [
              "service:action3",
              "service:action4"
            ],
            "effect": "Allow",
            "not_resources": [
              "resource1",
              "resource2"
            ]
          }
        ]
      }
    },
    "aws_partition": {
      "Partitition": {
        "provider": "aws"
      }
    }
  },
  "provider": {
    "aws": [
      {
        "region": "us-east-1"
      }
    ]
  },
  "terraform": {
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "5.68.0"
      }
    }
  }
}"
`;

exports[`IAM policy statement from JSON should not convert \`Principal: *\` to \`Principal: { AWS: * }\` 1`] = `
"{
  "data": {
    "aws_caller_identity": {
      "CallerIdentity": {
        "provider": "aws"
      }
    },
    "aws_iam_policy_document": {
      "doc_138423F6": {
        "statement": [
          {
            "actions": [
              "service:action1"
            ],
            "effect": "Allow",
            "principals": [
              {
                "identifiers": [
                  "*"
                ],
                "type": "*"
              }
            ],
            "resources": [
              "*"
            ]
          }
        ]
      }
    },
    "aws_partition": {
      "Partitition": {
        "provider": "aws"
      }
    }
  },
  "provider": {
    "aws": [
      {
        "region": "us-east-1"
      }
    ]
  },
  "terraform": {
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "5.68.0"
      }
    }
  }
}"
`;

exports[`IAM policy statement from JSON the kitchen sink 1`] = `
"{
  "data": {
    "aws_caller_identity": {
      "CallerIdentity": {
        "provider": "aws"
      }
    },
    "aws_iam_policy_document": {
      "doc_138423F6": {
        "statement": [
          {
            "actions": [
              "iam:ChangePassword"
            ],
            "effect": "Allow",
            "resources": [
              "*"
            ],
            "sid": "FirstStatement"
          },
          {
            "actions": [
              "s3:ListAllMyBuckets"
            ],
            "effect": "Allow",
            "resources": [
              "*"
            ],
            "sid": "SecondStatement"
          },
          {
            "actions": [
              "s3:List*",
              "s3:Get*"
            ],
            "condition": [
              {
                "test": "Bool",
                "values": [
                  "true"
                ],
                "variable": "aws:MultiFactorAuthPresent"
              }
            ],
            "effect": "Allow",
            "resources": [
              "arn:aws:s3:::confidential-data",
              "arn:aws:s3:::confidential-data/*"
            ],
            "sid": "ThirdStatement"
          }
        ]
      }
    },
    "aws_partition": {
      "Partitition": {
        "provider": "aws"
      }
    }
  },
  "provider": {
    "aws": [
      {
        "region": "us-east-1"
      }
    ]
  },
  "terraform": {
    "required_providers": {
      "aws": {
        "source": "aws",
        "version": "5.68.0"
      }
    }
  }
}"
`;
